### PR TITLE
[Feature] Support SOLO for dev-1.x

### PR DESCRIPTION
**Pull request by hanrui1sensetime**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Support Instance Segmentation model SOLO for dev-1.x

## Modification

- [x] Support SOLO model
- [ ] reg test
- [ ] docs and UT

Note: 
1. SOLO needs to change the `export_postprocess_mask = True` in `configs/mmdet/_base_/base_instance-seg_static.py` manually.
2. SOLO dynamic supports 320x320-416x416 inputs. Larger inputs may cause cuda allocation error.

## BC-breaking (Optional)

None.

## Use cases (Optional)

None.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: configs/mmdet/_base_/base_instance-seg_static_solo.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+_base_ = ['./base_static.py']
+
+onnx_config = dict(output_names=['dets', 'labels', 'masks'])
+codebase_config = dict(post_processing=dict(export_postprocess_mask=True))
```
Comment (grimoire) [inline configs/mmdet/_base_/base_instance-seg_static_solo.py:4]: Why do we have to export postprocess mask for solo only?

Comment (hanrui1sensetime) [reply configs/mmdet/_base_/base_instance-seg_static_solo.py:4]: > Why do we have to export postprocess mask for solo only?

SOLO calculate the mask directly, while other models generate masks from det bboxes, which is dummy in SOLO.

**Comment by RunningLeon** [comment]
@hanrui1sensetime have you tested the evaluation? could update the benchmark docs

**Comment by hanrui1sensetime** [comment]
> I can not test solo on my device. OOM.

Increase max_workspace_size to 1>>33, FYI.

File: configs/mmdet/_base_/base_instance-seg_dynamic_solo.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+_base_ = ['./base_instance-seg_static_solo.py']
```
Comment (RunningLeon) [inline configs/mmdet/_base_/base_instance-seg_dynamic_solo.py:1]: could remove these newly added base config

File: configs/mmdet/instance-seg/instance-seg_onnxruntime_dynamic_solo.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
+_base_ = [
```
Comment (RunningLeon) [inline configs/mmdet/instance-seg/instance-seg_onnxruntime_dynamic_solo.py:1]: better change to algo name after task, for instance `instance-seg_solo_onnxruntime_dynamic.py`

File: mmdeploy/codebase/mmdet/models/dense_heads/__init__.py [hunk 4]
```diff
@@ -7,6 +7,7 @@
 from .reppoints_head import reppoints_head__predict_by_feat
 from .rpn_head import rpn_head__get_bboxes__ncnn, rpn_head__predict_by_feat
 from .rtmdet_head import rtmdet_head__predict_by_feat
+from .solo_head import solohead__predict_by_feat__default
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmdet/models/dense_heads/__init__.py:10]: fix conflicts with latest dev-1.x and change the style of importing rewriting functions.

File: mmdeploy/codebase/mmdet/models/layers/__init__.py [hunk 5]
```diff
@@ -1,4 +1,5 @@
 # Copyright (c) OpenMMLab. All rights reserved.
 from .bbox_nms import multiclass_nms
+from .matrix_nms import mask_matrix_nms__default
```
Comment (grimoire) [inline mmdeploy/codebase/mmdet/models/layers/__init__.py:3]: `from . import matrix_nms`
We do not need to expose the rewriter.

File: mmdeploy/codebase/mmdet/models/detectors/single_stage.py [hunk 6]
```diff
@@ -74,3 +90,56 @@ def single_stage_detector__forward(ctx,
 
     return __forward_impl(
         ctx, self, batch_inputs, data_samples=data_samples, **kwargs)
+
+
+@FUNCTION_REWRITER.register_rewriter(
```
Comment (grimoire) [inline mmdeploy/codebase/mmdet/models/detectors/single_stage.py:95]: Should we move this rewriter to single_stage_instance_seg.py? @RunningLeon

File: configs/mmdet/instance-seg/instance-seg_solo_tensorrt-fp16_dynamic-320x320-1344x1344.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+_base_ = [
+    '../_base_/base_instance-seg_dynamic.py',
+    '../../_base_/backends/tensorrt-fp16.py'
+]
+
+codebase_config = dict(post_processing=dict(export_postprocess_mask=True))
+
+backend_config = dict(
+    common_config=dict(max_workspace_size=1 << 33),
```
Comment (grimoire) [inline configs/mmdet/instance-seg/instance-seg_solo_tensorrt-fp16_dynamic-320x320-1344x1344.py:9]: Errrr, I do not prefer too many configs.

File: docs/en/03-benchmark/benchmark.md [hunk 8]
```diff
@@ -823,26 +823,28 @@ Users can directly test the performance through [how_to_evaluate_a_model.md](../
     <td align="center">-</td>
   </tr>
   <tr>
-    <td align="center" rowspan="2"><a href="https://github.com/open-mmlab/mmdetection/tree/master/configs/mask_rcnn/mask_rcnn_r50_fpn_1x_coco.py">Mask R-CNN</a></td>
-    <td align="center" rowspan="2">Instance Segmentation</td>
-    <td align="center" rowspan="2">COCO2017</td>
+    <td align="center"><a href="https://github.com/open-mmlab/mmdetection/tree/master/configs/detr/detr_r50_8x2_150e_coco.py">DETR</a></td>
```
Comment (grimoire) [inline docs/en/03-benchmark/benchmark.md:826]: Why do we remove mask r-cnn and copy detr here?

File: configs/mmdet/instance-seg/instance-seg_solo_onnxruntime_dynamic.py [hunk 9]
```diff
@@ -0,0 +1,6 @@
+_base_ = [
```
Comment (grimoire) [inline configs/mmdet/instance-seg/instance-seg_solo_onnxruntime_dynamic.py:1]: I do not like the file name, different from what we have defined before.

**Review by grimoire — COMMENTED** [review]
Ort visualize failed with log:
```
2022-11-08 16:21:41.227929999 [E:onnxruntime:, sequential_executor.cc:339 Execute] Non-zero status code returned while running Concat node. Name:'Concat_910' Status Message: concat.cc:84 PrepareForCompute Non concat axis dimensions must match: Axis 2 has mismatched dimensions of 26 and 25
2022-11-08:16:21:41 - root - ERROR - Error in execution: Non-zero status code returned while running Concat node. Name:'Concat_910' Status Message: concat.cc:84 PrepareForCompute Non concat axis dimensions must match: Axis 2 has mismatched dimensions of 26 and 25

```

File: mmdeploy/codebase/mmdet/models/dense_heads/solo_head.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    batch_mlvl_mask_preds = torch.cat(mlvl_mask_preds, dim=1)
+    featmap_size = batch_mlvl_mask_preds.size()[-2:]
+    h, w = batch_img_metas[0]['img_shape'][:2]
+    upsampled_size = (featmap_size[0] * 4, featmap_size[1] * 4)
+    cls_labels = batch_mlvl_cls_scores.argmax(dim=-1)
+    batch_mlvl_cls_scores = torch.max(batch_mlvl_cls_scores, -1).values
+    score_mask = (batch_mlvl_cls_scores > cfg.score_thr)
+    batch_mlvl_cls_scores = batch_mlvl_cls_scores.where(
+        score_mask, batch_mlvl_cls_scores.new_zeros(1)).view(-1)
+    if len(batch_mlvl_cls_scores) == 0:
```
Comment (grimoire) [inline mmdeploy/codebase/mmdet/models/dense_heads/solo_head.py:46]: It is a bad idea to apply branch inside a rewriter, the trace will not get the expected result.

File: mmdeploy/codebase/mmdet/models/dense_heads/solo_head.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        return empty_results(batch_mlvl_cls_scores,
+                             batch_img_metas[0]['ori_shape'][:2])
+    cls_labels = cls_labels.view(-1)
+
+    # Filter the mask mask with an area is smaller than
+    # stride of corresponding feature level
+    batch_lvl_interval = cls_labels.new_tensor(self.num_grids).pow(2).cumsum(0)
+    strides = batch_mlvl_cls_scores.new_ones(batch_lvl_interval[-1])
+    strides[:batch_lvl_interval[0]] *= self.strides[0]
```
Comment (grimoire) [inline mmdeploy/codebase/mmdet/models/dense_heads/solo_head.py:55]: set_item is terrible when convert to onnx.

File: mmdeploy/codebase/mmdet/models/dense_heads/solo_head.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    # mask_matrix_nms may return an empty Tensor
+
+    if len(keep_inds) == 0:
+        return empty_results(cls_scores, batch_img_metas[0]['ori_shape'][:2])
+    mask_preds = mask_preds[keep_inds]
+    mask_preds = F.interpolate(
+        mask_preds.unsqueeze(0), size=upsampled_size,
+        mode='bilinear')[:, :, :h, :w]
+    labels = labels.reshape(batch_size, -1)
+    bboxes = scores.new_zeros(scores.shape[-1], 4).view(batch_size, -1, 4)
```
Comment (grimoire) [inline mmdeploy/codebase/mmdet/models/dense_heads/solo_head.py:98]: bboxes is all zero?

Comment (hanrui1sensetime) [reply mmdeploy/codebase/mmdet/models/dense_heads/solo_head.py:98]: This bbox is dummy.

**Comment by RunningLeon** [comment]
@hanrui1sensetime may fix conflicts

**Comment by RunningLeon** [comment]
@hanrui1sensetime Hi, could fix conflicts at first.
